### ISSUE TITLE
npm install githubuser/repo shouldn't prefer git:// over HTTPS or SSH

### ISSUE THREAD

**Issue by cvrebert** [npm/npm#9470]
Using npm 2.13.1 (I didn't see anything relevant in the 2.14 changelog), installing a dependency specified using the `user/repo` GitHub shorthand tries to `git clone` from the following, in order:
1. `git://github.com/user/repo.git`
2. `https://github.com/user/repo.git`
3. `git@github.com:user/repo.git`

Since GitHub repos _always_ have HTTPS available and since `git://` is (a) not authenticated [oh hai, i'm in ur networkz, feeding you malicious codez] and (b) blocked by some (admittedly idiotic) firewalls, perhaps npm should skip `git://` altogether in this case?

**Comment by zkat** [comment]
After talking about it, we noted that github does default to `https://` for both reads and writes in its UI, and it's quite possible most people use that as their preference for git repos -- as such, we agree that it makes sense to make `https://` the default.

That said, removing `git://` altogether can (and probably will) move enough people's cheese (*specially* those users who have a harder time setting up network/protocol stuff, such as folks behind those annoying firewalls you mention. We think it's a good idea to move `git://` to be the last option, but we won't remove it for now. Your concerns about MITM attacks are noted, but we're not sure they have as big of an impact as the support issues with proxies.

Note that, if you want more reliable protection against malicious code getting installed, you might consider using the npm registry instead, which verifies checksums.

**Comment by cvrebert** [comment]
Sounds good!

**Comment by TheJeremyGray** [comment]
This will resolve my (idiotic?) issues around firewalls

**Comment by gagern** [comment]
Since various people seem to prefer various orders for various reasons, would it make sense to allow configuring the order, with a general setting for all services and an option to override that per service? That still leaves us with the task of defining sane defaults, but it would allow corner cases a way out.

**Comment by othiym23** [comment]
> would it make sense to allow configuring the order, with a general setting for all services and an option to override that per service? 

I believe that's too much complexity for the feature, which is "make it easy to install from hosted git services, without forcing people to think too hard about how it's done". It's possible, via a combination of git configuration (`.insteadOf`) and using explicit repo URLs in `package.json`, for people who need to deal with firewalls and proxies. Pushing that complexity into npm would mostly make it harder to test and support.

**Comment by othiym23** [comment]
As discussed on PR #11087, the canonical order should be `https:` → `ssh:` → `git:`, and we would welcome a patch to that effect (say, a reworked version of #11087, which was _really close_! or build on the start that @iarna made with https://github.com/npm/npm/compare/iarna/fixup-shortcut-order, adding tests and tweaking the documentation). Since better parity between git and registry dependencies is one of the project's goals, we'll get to this eventually, but it will probably be later rather than sooner given our current set or priorities.

**Comment by gagern** [comment]
@othiym23 See PR #11248. I changed the order as requested. Like @iarna I now have an array to define the order, but in my case it's an array of method names from the hosted service description, not an array of functions in the local module. Which means I don't need separate functions for the various protocols any more. I've adjusted test cases and documentation, and in the latter mentioned the `url.<base>.insteadOf` git configuration you suggested, since [the comment above](https://github.com/npm/npm/issues/9470#issuecomment-159601060) by @TheJeremyGray sounds as if there might be use cases for it.
